amm/interp/src/main/scala/ammonite/interp/Preprocessor.scala

```diff
@@ -297,35 +299,207 @@ object Preprocessor{
+package ${pkgName.head.encoded}
+package ${Util.encodeScalaSourcePath(pkgName.tail)}
 
-    def bottom(printCode: String, indexedWrapperName: Name, extraCode: String) =
-      normalizeNewlines(s"""\ndef $$main() = { $printCode }
-  override def toString = "${indexedWrapperName.raw}"
+$imports
+
+object ${indexedWrapperName.backticked}{
+  val instance: Helper.type = Helper
+  def $$main() = instance.$$main()
+
+  object Helper extends java.io.Serializable {
+"""
+        )
+
+        val bottom = normalizeNewlines(s"""\ndef $$main() = { $printCode }
+  override def toString = "${indexedWrapperName.encoded}"
   $extraCode
+}}
+""")
+
+        (top, bottom, userCodeNestingLevel)
+      } else {
+
+        val (reworkedImports, reqVals) = {
+
+          val (l, reqVals0) = imports
+            .value
+            .map { data =>
+              val prefix = Seq(Name("_root_"), Name("ammonite"), Name("$sess"))
+              if (data.prefix.startsWith(prefix) && data.prefix.endsWith(wrapperPath)) {
+                val name = data.prefix.drop(prefix.length).dropRight(wrapperPath.length).last
+                (data.copy(prefix = Seq(name)), Seq(name -> data.prefix))
+              } else
+                (data, Nil)
+            }
+            .unzip
+
+          (Imports(l), reqVals0.flatten)
+        }
+
+        val requiredVals = reqVals
+          .distinct
+          .groupBy(_._1)
+          .mapValues(_.map(_._2))
+          .toVector
+          .sortBy(_._1.raw)
+          .collect {
+            case (key, Seq(path)) =>
+              /*
+               * Via __amm_usedThings, that itself relies on the *-tree.txt resources generated
+               * via the AmmonitePlugin, we can know whether the current command uses things from
+               * each of the previous ones, and null-ify the references to those that are unused.
+               * That way, the unused commands don't prevent serializing this command.
+               */
+              val encoded = Util.encodeScalaSourcePath(path)
+              s"final val ${key.backticked}: $encoded.type = " +
+                s"if (__amm_usedThings($q$q$q${key.raw}$q$q$q)) $encoded else null$newLine"
+            case (key, values) =>
+              throw new Exception(
+                "Should not happen - several required values with the same name " +
+                  s"(name: $key, values: $values)"
+              )
+          }
+          .mkString
+
+        val top = normalizeNewlines(s"""
+package ${pkgName.head.encoded}
+package ${Util.encodeScalaSourcePath(pkgName.tail)}
+
+object ${indexedWrapperName.backticked}{
+  val wrapper = new ${indexedWrapperName.backticked}
+  val instance = new wrapper.Helper
+  def $$main() = instance.$$main()
 }
+
+final class ${indexedWrapperName.backticked} extends java.io.Serializable {
+
+  @_root_.scala.transient private val __amm_usedThings =
+    new _root_.ammonite.interp.Preprocessor.CodeClassWrapper.UsedThings(this)
```

Comment:
Rather than serializing the `usedThings` data into a classpath file which gets loaded at runtime and deserialized, can we just expose the structured data directly in a `repl.usedEarlierDefinitions` property? Then your conditionals can just be 

```scala
if (repl.usedEarlierDefinitions.contains("foo")) ... else null
```

`repl` already exposes `repl.imports` and `repl.history`, which depend on the current command and the exact order of commands that have run so far, so exposing a `repl.usedEarlierDefinitions` will fit right in

Reply:
Will do, that's better this way!

Reply:
@lihaoyi Wouldn't something like having the `usedEarlierDefinitions` rely on the resource be ok? I tried to pass the used commands everywhere around, it was quite heavyweight (I had to add parameters for it to a bunch of methods, plus it has to be cached then - so change the cache api, …)

Reply:
The advantage of the resource approach, is that they are passed along with the generated classes, so only require little changes.